server/src/proxy/grpc/write.rs

```diff
@@ -89,11 +98,47 @@ impl<Q: QueryExecutor + 'static> Proxy<Q> {
         ctx: Context,
         req: WriteRequest,
     ) -> Result<WriteResponse> {
+        let database = req
+            .context
+            .clone()
+            .context(ErrNoCause {
+                msg: "Missing context",
+                code: StatusCode::BAD_REQUEST,
+            })?
+            .database;
+        let (write_request_to_local, mut futures) = self.maybe_forward_write(req).await?;
+
+        // Write to local.
+        if !write_request_to_local.table_requests.is_empty() {
+            let local_future =
+                async move { self.write_to_local(ctx, write_request_to_local).await };
+            futures.push(local_future.boxed());
+        }
```

Comment:
How about splitting the `maybe_forward_write` into two parts: `split_requests` and `forward_requests`? After that, the `futures` capacity can be determined before use.